src/target/riscv/riscv.c

```diff
@@ -2372,12 +2356,26 @@ int riscv_openocd_step(struct target *target, int current,
 	if (disable_triggers(target, trigger_state) != ERROR_OK)
 		return ERROR_FAIL;
 
+	uint64_t current_mstatus;
+	RISCV_INFO(info);
+	if (info->isrmask_mode == RISCV_ISRMASK_STEPONLY) {
+		/* Disable Interrupts before stepping. */
+		uint64_t ie_mask = MSTATUS_MIE | MSTATUS_HIE | MSTATUS_SIE | MSTATUS_UIE;
+		if (riscv_interrupts_disable(target, ie_mask, &current_mstatus) != ERROR_OK)
+			return ERROR_FAIL;
+	}
+
 	int out = riscv_step_rtos_hart(target);
 	if (out != ERROR_OK) {
 		LOG_ERROR("unable to step rtos hart");
 		return out;
 	}
 
+	if (info->isrmask_mode == RISCV_ISRMASK_STEPONLY) {
+		if (riscv_interrupts_restore(target, current_mstatus) != ERROR_OK)
+			return ERROR_FAIL;
+	}
+
```

Comment:
@JanMatCodasip thanks to point out this behavior. Looking at the other parts,  I was in between restore/return or just return. So what about enabling triggers if there is step or irq_restore  error?

Reply:
@timsifive your comments are also welcome.

Reply:
> Looking at the other parts, I was in between restore/return or just return.

You're right - the risk of not restoring everything has already been present in the original code (restore of triggers would not be attempted if step fails).

> So what about enabling triggers if there is step or irq_restore error?

IMO, an ideal solution would be to rework this function so that in case of any failure in any of the steps, whatever got changed gets restored.  I understand that this is additional work, not directly related to the new "maskisr" command.

Please create an issue for that so that we don't forget to return to this.

Reply:
Didn't want to create another technical debt.  I did some restore work. Please check it again.